Quarkus CLI finds and uses rogue application.properties files and crashes

### Describe the bug

I recently installed the Quarkus CLI and upon first execution of `quarkus version` it spit out tons of warnings and I had no idea why (I was not in any project directory, rather was in my home directory where I never put projects). Turns out I had a rogue `application.properties` in the directory (with no other project-related files) and it loaded and parsed it as part of the CLI runtime.

If that file is a binary (e.g. by renaming a .zip file to `application.properties`) then Quarkus CLI crashes.

### Expected behavior

No warnings on first execution, in a directory with no project.

### Actual behavior

With a valid properties file:

```sh
➜  ~ quarkus version
2021-08-20 12:48:03,173 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.hibernate-orm.sql-load-script" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-08-20 12:48:03,173 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.datasource.username" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-08-20 12:48:03,173 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.datasource.jdbc.url" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-08-20 12:48:03,174 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.hibernate-orm.database.generation" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-08-20 12:48:03,174 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.datasource.db-kind" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
2021-08-20 12:48:03,174 WARN  [io.qua.config] (main) Unrecognized configuration key "quarkus.datasource.password" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo
Client Version 2.1.2.Final
```
With a binary named `application.properties`:
```sh
➜  ~ cp v3.11.0.zip application.properties
➜  ~ quarkus version
Exception in thread "main" java.lang.ExceptionInInitializerError
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:60)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:65)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:30)
Caused by: java.lang.IllegalArgumentException: Malformed \uxxxx encoding.
```



### How to Reproduce?

1. Take any file and call it `application.properties`
1. Install CLI with: `curl -Ls https://sh.jbang.dev | bash -s - app install --fresh --force quarkus@quarkusio`
1. While in the same directory as the rogue file, run `quarkus version`

### Output of `uname -a` or `ver`

Darwin jfalkner-OSX 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64 x86_64

### Output of `java -version`

```
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
```

### GraalVM version (if different from Java)

N/A

### Quarkus version or git rev

Client Version 2.1.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Quarkus CLI

### Additional information

I'm concerned it may be possible to do something nasty with a cleverly-designed `application.properties`.

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -55,7 +55,7 @@
         <smallrye-reactive-types-converter.version>2.6.0</smallrye-reactive-types-converter.version>
         <smallrye-mutiny-vertx-binding.version>2.15.1</smallrye-mutiny-vertx-binding.version>
         <smallrye-reactive-messaging.version>3.11.0</smallrye-reactive-messaging.version>
-        <smallrye-stork.version>1.0.0.Beta1</smallrye-stork.version>    
+        <smallrye-stork.version>1.0.0.Beta1</smallrye-stork.version>
         <jakarta.activation.version>1.2.1</jakarta.activation.version>
         <jakarta.annotation-api.version>1.3.5</jakarta.annotation-api.version>
         <jakarta.el-impl.version>3.0.4</jakarta.el-impl.version>

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/RunTimeConfigBuilderBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/RunTimeConfigBuilderBuildItem.java
@@ -1,0 +1,15 @@
+package io.quarkus.deployment.builditem;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class RunTimeConfigBuilderBuildItem extends MultiBuildItem {
+    private String builderClassName;
+
+    public RunTimeConfigBuilderBuildItem(final String builderClassName) {
+        this.builderClassName = builderClassName;
+    }
+
+    public String getBuilderClassName() {
+        return builderClassName;
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/StaticInitConfigBuilderBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/StaticInitConfigBuilderBuildItem.java
@@ -1,0 +1,15 @@
+package io.quarkus.deployment.builditem;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class StaticInitConfigBuilderBuildItem extends MultiBuildItem {
+    private String builderClassName;
+
+    public StaticInitConfigBuilderBuildItem(final String builderClassName) {
+        this.builderClassName = builderClassName;
+    }
+
+    public String getBuilderClassName() {
+        return builderClassName;
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java
@@ -173,6 +173,8 @@ public final class RunTimeConfigurationGenerator {
             ConfigUtils.class,
             "configBuilder",
             SmallRyeConfigBuilder.class, boolean.class, boolean.class, boolean.class, LaunchMode.class);
+    static final MethodDescriptor CU_CONFIG_BUILDER_LIST = MethodDescriptor.ofMethod(ConfigUtils.class, "configBuilder",
+            SmallRyeConfigBuilder.class, SmallRyeConfigBuilder.class, List.class);
     static final MethodDescriptor CU_ADD_SOURCE_PROVIDER = MethodDescriptor.ofMethod(ConfigUtils.class, "addSourceProvider",
             void.class, SmallRyeConfigBuilder.class, ConfigSourceProvider.class);
     static final MethodDescriptor CU_ADD_SOURCE_PROVIDERS = MethodDescriptor.ofMethod(ConfigUtils.class, "addSourceProviders",
@@ -304,11 +306,13 @@ public final class RunTimeConfigurationGenerator {
         final Set<String> staticConfigSources;
         final Set<String> staticConfigSourceProviders;
         final Set<String> staticConfigSourceFactories;
+        final Set<String> staticConfigBuilders;
         final Set<String> runtimeConfigSources;
         final Set<String> runtimeConfigSourceProviders;
         final Set<String> runtimeConfigSourceFactories;
         final Set<ConfigClassWithPrefix> staticConfigMappings;
         final Set<ConfigClassWithPrefix> runtimeConfigMappings;
+        final Set<String> runtimeConfigBuilders;
         /**
          * Regular converters organized by type. Each converter is stored in a separate field. Some are used
          * only at build time, some only at run time, and some at both times.
@@ -342,11 +346,13 @@ public final class RunTimeConfigurationGenerator {
             staticConfigSources = builder.getStaticConfigSources();
             staticConfigSourceProviders = builder.getStaticConfigSourceProviders();
             staticConfigSourceFactories = builder.getStaticConfigSourceFactories();
+            staticConfigBuilders = builder.getStaticConfigBuilders();
             runtimeConfigSources = builder.getRuntimeConfigSources();
             runtimeConfigSourceProviders = builder.getRuntimeConfigSourceProviders();
             runtimeConfigSourceFactories = builder.getRuntimeConfigSourceFactories();
             staticConfigMappings = builder.getStaticConfigMappings();
             runtimeConfigMappings = builder.getRuntimeConfigMappings();
+            runtimeConfigBuilders = builder.getRuntimeConfigBuilders();
             cc = ClassCreator.builder().classOutput(classOutput).className(CONFIG_CLASS_NAME).setFinal(true).build();
             generateEmptyParsers(cc);
             // not instantiable
@@ -439,6 +445,14 @@ public final class RunTimeConfigurationGenerator {
                 clinit.invokeStaticMethod(CU_WITH_MAPPING, buildTimeBuilder,
                         clinit.load(configMapping.getKlass().getName()), clinit.load(configMapping.getPrefix()));
             }
+
+            // additional config builders
+            ResultHandle staticConfigBuilderInstances = clinit.newInstance(AL_NEW);
+            for (String configBuilder : staticConfigBuilders) {
+                ResultHandle staticConfigBuilderInstance = clinit.newInstance(MethodDescriptor.ofConstructor(configBuilder));
+                clinit.invokeVirtualMethod(AL_ADD, staticConfigBuilderInstances, staticConfigBuilderInstance);
+            }
+            clinit.invokeStaticMethod(CU_CONFIG_BUILDER_LIST, buildTimeBuilder, staticConfigBuilderInstances);
 
             clinitConfig = clinit.checkCast(clinit.invokeVirtualMethod(SRCB_BUILD, buildTimeBuilder),
                     SmallRyeConfig.class);
@@ -699,6 +713,15 @@ public final class RunTimeConfigurationGenerator {
                 readConfig.invokeStaticMethod(CU_WITH_MAPPING, runTimeBuilder,
                         readConfig.load(configMapping.getKlass().getName()), readConfig.load(configMapping.getPrefix()));
             }
+
+            // additional config builders
+            ResultHandle runtimeConfigBuilderInstances = readConfig.newInstance(AL_NEW);
+            for (String configBuilder : runtimeConfigBuilders) {
+                ResultHandle runtimeConfigBuilderInstance = readConfig
+                        .newInstance(MethodDescriptor.ofConstructor(configBuilder));
+                readConfig.invokeVirtualMethod(AL_ADD, runtimeConfigBuilderInstances, runtimeConfigBuilderInstance);
+            }
+            readConfig.invokeStaticMethod(CU_CONFIG_BUILDER_LIST, runTimeBuilder, runtimeConfigBuilderInstances);
 
             ResultHandle bootstrapConfig = null;
             if (bootstrapConfigSetupNeeded()) {
@@ -1712,9 +1735,11 @@ public final class RunTimeConfigurationGenerator {
             private Set<String> staticConfigSources;
             private Set<String> staticConfigSourceProviders;
             private Set<String> staticConfigSourceFactories;
+            private Set<String> staticConfigBuilders;
             private Set<String> runtimeConfigSources;
             private Set<String> runtimeConfigSourceProviders;
             private Set<String> runtimeConfigSourceFactories;
+            private Set<String> runtimeConfigBuilders;
 
             private Set<ConfigClassWithPrefix> staticConfigMappings;
             private Set<ConfigClassWithPrefix> runtimeConfigMappings;
@@ -1799,6 +1824,15 @@ public final class RunTimeConfigurationGenerator {
                 return this;
             }
 
+            Set<String> getStaticConfigBuilders() {
+                return staticConfigBuilders;
+            }
+
+            public Builder setStaticConfigBuilders(final Set<String> staticConfigBuilders) {
+                this.staticConfigBuilders = staticConfigBuilders;
+                return this;
+            }
+
             Set<String> getRuntimeConfigSources() {
                 return runtimeConfigSources;
             }
@@ -1841,6 +1875,15 @@ public final class RunTimeConfigurationGenerator {
 
             public Builder setRuntimeConfigMappings(final Set<ConfigClassWithPrefix> runtimeConfigMappings) {
                 this.runtimeConfigMappings = runtimeConfigMappings;
+                return this;
+            }
+
+            Set<String> getRuntimeConfigBuilders() {
+                return runtimeConfigBuilders;
+            }
+
+            public Builder setRuntimeConfigBuilders(final Set<String> runtimeConfigBuilders) {
+                this.runtimeConfigBuilders = runtimeConfigBuilders;
                 return this;
             }
 

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
@@ -40,7 +40,9 @@ import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.LiveReloadBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigBuilderBuildItem;
 import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.StaticInitConfigBuilderBuildItem;
 import io.quarkus.deployment.builditem.StaticInitConfigSourceFactoryBuildItem;
 import io.quarkus.deployment.builditem.StaticInitConfigSourceProviderBuildItem;
 import io.quarkus.deployment.builditem.SuppressNonRuntimeConfigChangedWarningBuildItem;
@@ -113,7 +115,9 @@ public class ConfigGenerationBuildStep {
             List<AdditionalBootstrapConfigSourceProviderBuildItem> additionalBootstrapConfigSourceProviders,
             List<StaticInitConfigSourceProviderBuildItem> staticInitConfigSourceProviders,
             List<StaticInitConfigSourceFactoryBuildItem> staticInitConfigSourceFactories,
-            List<ConfigMappingBuildItem> configMappings)
+            List<ConfigMappingBuildItem> configMappings,
+            List<StaticInitConfigBuilderBuildItem> staticInitConfigBuilders,
+            List<RunTimeConfigBuilderBuildItem> runTimeConfigBuilders)
             throws IOException {
 
         if (liveReloadBuildItem.isLiveReload()) {
@@ -127,7 +131,7 @@ public class ConfigGenerationBuildStep {
         Set<String> staticConfigSourceProviders = new HashSet<>();
         staticConfigSourceProviders.addAll(staticSafeServices(discoveredConfigSourceProviders));
         staticConfigSourceProviders.addAll(staticInitConfigSourceProviders.stream()
-                .map(StaticInitConfigSourceProviderBuildItem::getProviderClassName).collect(Collectors.toSet()));
+                .map(StaticInitConfigSourceProviderBuildItem::getProviderClassName).collect(toSet()));
         Set<String> staticConfigSourceFactories = new HashSet<>();
         staticConfigSourceFactories.addAll(staticSafeServices(discoveredConfigSourceFactories));
         staticConfigSourceFactories.addAll(staticInitConfigSourceFactories.stream()
@@ -146,11 +150,15 @@ public class ConfigGenerationBuildStep {
                 .setStaticConfigSources(staticSafeServices(discoveredConfigSources))
                 .setStaticConfigSourceProviders(staticConfigSourceProviders)
                 .setStaticConfigSourceFactories(staticConfigSourceFactories)
+                .setStaticConfigMappings(staticSafeConfigMappings(configMappings))
+                .setStaticConfigBuilders(staticInitConfigBuilders.stream()
+                        .map(StaticInitConfigBuilderBuildItem::getBuilderClassName).collect(toSet()))
                 .setRuntimeConfigSources(discoveredConfigSources)
                 .setRuntimeConfigSourceProviders(discoveredConfigSourceProviders)
                 .setRuntimeConfigSourceFactories(discoveredConfigSourceFactories)
-                .setStaticConfigMappings(staticSafeConfigMappings(configMappings))
                 .setRuntimeConfigMappings(runtimeConfigMappings(configMappings))
+                .setRuntimeConfigBuilders(
+                        runTimeConfigBuilders.stream().map(RunTimeConfigBuilderBuildItem::getBuilderClassName).collect(toSet()))
                 .build()
                 .run();
     }

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigBuilder.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigBuilder.java
@@ -1,0 +1,11 @@
+package io.quarkus.runtime.configuration;
+
+import io.smallrye.config.SmallRyeConfigBuilder;
+
+public interface ConfigBuilder {
+    SmallRyeConfigBuilder configBuilder(SmallRyeConfigBuilder builder);
+
+    default int priority() {
+        return 0;
+    }
+}

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java
@@ -1,6 +1,5 @@
 package io.quarkus.runtime.configuration;
 
-import static io.smallrye.config.DotEnvConfigSourceProvider.dotEnvSources;
 import static io.smallrye.config.PropertiesConfigSourceProvider.classPathSources;
 import static io.smallrye.config.SmallRyeConfig.SMALLRYE_CONFIG_LOCATIONS;
 import static io.smallrye.config.SmallRyeConfig.SMALLRYE_CONFIG_PROFILE;
@@ -13,6 +12,7 @@ import java.net.URL;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
@@ -70,7 +70,7 @@ public final class ConfigUtils {
         return size -> new TreeSet<>();
     }
 
-    public static SmallRyeConfigBuilder configBuilder(final boolean runTime, LaunchMode launchMode) {
+    public static SmallRyeConfigBuilder configBuilder(final boolean runTime, final LaunchMode launchMode) {
         return configBuilder(runTime, true, launchMode);
     }
 
@@ -82,7 +82,7 @@ public final class ConfigUtils {
      * @return the configuration builder
      */
     public static SmallRyeConfigBuilder configBuilder(final boolean runTime, final boolean addDiscovered,
-            LaunchMode launchMode) {
+            final LaunchMode launchMode) {
         return configBuilder(runTime, false, addDiscovered, launchMode);
     }
 
@@ -94,22 +94,22 @@ public final class ConfigUtils {
      * @return the configuration builder
      */
     public static SmallRyeConfigBuilder configBuilder(final boolean runTime, final boolean bootstrap,
-            final boolean addDiscovered,
-            LaunchMode launchMode) {
-        final SmallRyeConfigBuilder builder = emptyConfigBuilder();
+            final boolean addDiscovered, final LaunchMode launchMode) {
+        SmallRyeConfigBuilder builder = emptyConfigBuilder();
 
-        final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-        builder.withSources(new ApplicationPropertiesConfigSourceLoader.InFileSystem().getConfigSources(classLoader));
-        builder.withSources(new ApplicationPropertiesConfigSourceLoader.InClassPath().getConfigSources(classLoader));
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        builder.forClassLoader(classLoader);
+        builder.withSources(new ApplicationPropertiesConfigSourceLoader.InFileSystem());
+        builder.withSources(new ApplicationPropertiesConfigSourceLoader.InClassPath());
         if (launchMode.isDevOrTest() && (runTime || bootstrap)) {
             builder.withSources(new RuntimeOverrideConfigSource(classLoader));
         }
         if (runTime) {
             builder.addDefaultSources();
             builder.withDefaultValue(UUID_KEY, UUID.randomUUID().toString());
-            builder.withSources(dotEnvSources(classLoader));
+            builder.withSources(new DotEnvConfigSourceProvider());
         } else {
-            final List<ConfigSource> sources = new ArrayList<>();
+            List<ConfigSource> sources = new ArrayList<>();
             sources.addAll(classPathSources(META_INF_MICROPROFILE_CONFIG_PROPERTIES, classLoader));
             sources.addAll(new BuildTimeDotEnvConfigSourceProvider().getConfigSources(classLoader));
             sources.add(new BuildTimeEnvConfigSource());
@@ -127,10 +127,10 @@ public final class ConfigUtils {
     }
 
     public static SmallRyeConfigBuilder emptyConfigBuilder() {
-        final SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
+        SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
         builder.withDefaultValue(SMALLRYE_CONFIG_PROFILE, ProfileManager.getActiveProfile());
 
-        final Map<String, String> relocations = new HashMap<>();
+        Map<String, String> relocations = new HashMap<>();
         relocations.put(SMALLRYE_CONFIG_LOCATIONS, "quarkus.config.locations");
         relocations.put(SMALLRYE_CONFIG_PROFILE_PARENT, "quarkus.config.profile.parent");
         // Override the priority, because of the ProfileConfigSourceInterceptor and profile.parent.
@@ -146,7 +146,7 @@ public final class ConfigUtils {
             }
         });
 
-        final Map<String, String> fallbacks = new HashMap<>();
+        Map<String, String> fallbacks = new HashMap<>();
         fallbacks.put("quarkus.config.locations", SMALLRYE_CONFIG_LOCATIONS);
         fallbacks.put("quarkus.config.profile.parent", SMALLRYE_CONFIG_PROFILE_PARENT);
         builder.withInterceptorFactories(new ConfigSourceInterceptorFactory() {
@@ -165,6 +165,20 @@ public final class ConfigUtils {
         builder.addDiscoveredInterceptors();
         builder.addDiscoveredConverters();
         builder.addDiscoveredValidator();
+        return builder;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static SmallRyeConfigBuilder configBuilder(SmallRyeConfigBuilder builder, List<ConfigBuilder> configBuilders) {
+        configBuilders.sort(Comparator.comparing(ConfigBuilder::priority));
+
+        for (ConfigBuilder configBuilder : configBuilders) {
+            builder = configBuilder.configBuilder(builder);
+            if (builder == null) {
+                throw new ConfigurationException(configBuilder.getClass().getName() + " returned a null builder");
+            }
+        }
+
         return builder;
     }
 

--- a/core/test-extension/deployment/src/main/java/io/quarkus/extest/deployment/TestProcessor.java
+++ b/core/test-extension/deployment/src/main/java/io/quarkus/extest/deployment/TestProcessor.java
@@ -38,8 +38,10 @@ import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.LogHandlerBuildItem;
 import io.quarkus.deployment.builditem.ObjectSubstitutionBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigBuilderBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
+import io.quarkus.deployment.builditem.StaticInitConfigBuilderBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.NativeImageResourceBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.RuntimeInitializedPackageBuildItem;
@@ -51,6 +53,8 @@ import io.quarkus.extest.runtime.beans.CommandServlet;
 import io.quarkus.extest.runtime.beans.PublicKeyProducer;
 import io.quarkus.extest.runtime.config.ObjectOfValue;
 import io.quarkus.extest.runtime.config.ObjectValueOf;
+import io.quarkus.extest.runtime.config.RunTimeConfigBuilder;
+import io.quarkus.extest.runtime.config.StaticInitConfigBuilder;
 import io.quarkus.extest.runtime.config.TestBuildAndRunTimeConfig;
 import io.quarkus.extest.runtime.config.TestBuildTimeConfig;
 import io.quarkus.extest.runtime.config.TestConfigRoot;
@@ -406,6 +410,16 @@ public final class TestProcessor {
             BuildProducer<AdditionalStaticInitConfigSourceProviderBuildItem> additionalStaticInitConfigSourceProviders) {
         additionalStaticInitConfigSourceProviders.produce(new AdditionalStaticInitConfigSourceProviderBuildItem(
                 AdditionalStaticInitConfigSourceProvider.class.getName()));
+    }
+
+    @BuildStep
+    void staticInitConfigBuilder(BuildProducer<StaticInitConfigBuilderBuildItem> configBuilders) {
+        configBuilders.produce(new StaticInitConfigBuilderBuildItem(StaticInitConfigBuilder.class.getName()));
+    }
+
+    @BuildStep
+    void runTimeConfigBuilder(BuildProducer<RunTimeConfigBuilderBuildItem> configBuilders) {
+        configBuilders.produce(new RunTimeConfigBuilderBuildItem(RunTimeConfigBuilder.class.getName()));
     }
 
     @BuildStep(onlyIf = Never.class)

--- a/core/test-extension/deployment/src/test/java/io/quarkus/config/ConfigBuilderTest.java
+++ b/core/test-extension/deployment/src/test/java/io/quarkus/config/ConfigBuilderTest.java
@@ -1,0 +1,33 @@
+package io.quarkus.config;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.extest.runtime.config.StaticInitConfigBuilder;
+import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.config.SmallRyeConfig;
+
+public class ConfigBuilderTest {
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class));
+
+    @Inject
+    SmallRyeConfig config;
+
+    @Test
+    void staticConfigBuilder() {
+        assertEquals(1, StaticInitConfigBuilder.counter.get());
+    }
+
+    @Test
+    void runTimeConfigBuilder() {
+        assertEquals("1234", config.getRawValue("additional.builder.property"));
+    }
+}

--- a/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/config/RunTimeConfigBuilder.java
+++ b/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/config/RunTimeConfigBuilder.java
@@ -1,0 +1,12 @@
+package io.quarkus.extest.runtime.config;
+
+import io.quarkus.runtime.configuration.ConfigBuilder;
+import io.smallrye.config.SmallRyeConfigBuilder;
+
+public class RunTimeConfigBuilder implements ConfigBuilder {
+    @Override
+    public SmallRyeConfigBuilder configBuilder(final SmallRyeConfigBuilder builder) {
+        builder.withDefaultValue("additional.builder.property", "1234");
+        return builder;
+    }
+}

--- a/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/config/StaticInitConfigBuilder.java
+++ b/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/config/StaticInitConfigBuilder.java
@@ -1,0 +1,19 @@
+package io.quarkus.extest.runtime.config;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.quarkus.runtime.configuration.ConfigBuilder;
+import io.smallrye.config.SmallRyeConfigBuilder;
+
+public class StaticInitConfigBuilder implements ConfigBuilder {
+    public static AtomicInteger counter = new AtomicInteger(0);
+
+    public StaticInitConfigBuilder() {
+        counter.incrementAndGet();
+    }
+
+    @Override
+    public SmallRyeConfigBuilder configBuilder(final SmallRyeConfigBuilder builder) {
+        return builder;
+    }
+}

--- a/extensions/picocli/deployment/pom.xml
+++ b/extensions/picocli/deployment/pom.xml
@@ -25,6 +25,18 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-picocli</artifactId>
         </dependency>
+
+        <!-- test dependencies -->
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5-internal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/picocli/deployment/src/main/java/io/quarkus/picocli/deployment/PicocliDeploymentConfiguration.java
+++ b/extensions/picocli/deployment/src/main/java/io/quarkus/picocli/deployment/PicocliDeploymentConfiguration.java
@@ -26,4 +26,13 @@ class PicocliDeploymentConfiguration {
      */
     @ConfigItem(name = "native-image.processing.enable", defaultValue = "true")
     boolean nativeImageProcessingEnabled;
+
+    /**
+     * <p>
+     * Set this to <code>true</code> to disable reading configuration from the default Config Sources. This includes
+     * `application.properties` and `.env` files.
+     * </p>
+     */
+    @ConfigItem(defaultValue = "false")
+    boolean disableDefaultConfigSources;
 }

--- a/extensions/picocli/deployment/src/main/java/io/quarkus/picocli/deployment/PicocliProcessor.java
+++ b/extensions/picocli/deployment/src/main/java/io/quarkus/picocli/deployment/PicocliProcessor.java
@@ -21,7 +21,10 @@ import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.IndexDependencyBuildItem;
 import io.quarkus.deployment.builditem.QuarkusApplicationClassBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigBuilderBuildItem;
+import io.quarkus.deployment.builditem.StaticInitConfigBuilderBuildItem;
 import io.quarkus.picocli.runtime.DefaultPicocliCommandLineFactory;
+import io.quarkus.picocli.runtime.PicocliConfigBuilder;
 import io.quarkus.picocli.runtime.PicocliRunner;
 import io.quarkus.picocli.runtime.annotations.TopCommand;
 import io.quarkus.runtime.annotations.QuarkusMain;
@@ -42,6 +45,17 @@ class PicocliProcessor {
     @BuildStep
     IndexDependencyBuildItem picocliIndexDependency() {
         return new IndexDependencyBuildItem("info.picocli", "picocli");
+    }
+
+    @BuildStep
+    void configBuilder(
+            PicocliDeploymentConfiguration configuration,
+            BuildProducer<StaticInitConfigBuilderBuildItem> staticInitConfigBuilder,
+            BuildProducer<RunTimeConfigBuilderBuildItem> runTimeConfigBuilder) {
+        if (configuration.disableDefaultConfigSources) {
+            staticInitConfigBuilder.produce(new StaticInitConfigBuilderBuildItem(PicocliConfigBuilder.class.getName()));
+            runTimeConfigBuilder.produce(new RunTimeConfigBuilderBuildItem(PicocliConfigBuilder.class.getName()));
+        }
     }
 
     @BuildStep

--- a/extensions/picocli/deployment/src/test/java/io/quarkus/picocli/deployment/AvailableConfigSourcesTest.java
+++ b/extensions/picocli/deployment/src/test/java/io/quarkus/picocli/deployment/AvailableConfigSourcesTest.java
@@ -1,0 +1,38 @@
+package io.quarkus.picocli.deployment;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.config.ConfigValue;
+import io.smallrye.config.SmallRyeConfig;
+
+public class AvailableConfigSourcesTest {
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest().withApplicationRoot(
+            (jar) -> jar.addAsResource(new StringAsset("quarkus.picocli.disable-default-config-sources=true\n" +
+                    "my.prop=1234\n"), "application.properties"));
+
+    @Inject
+    SmallRyeConfig config;
+
+    @Test
+    void sources() {
+        ConfigValue value = config.getConfigValue("my.prop");
+        assertEquals("1234", value.getValue());
+        assertEquals("PropertiesConfigSource[source=Specified default values]", value.getConfigSourceName());
+
+        for (final ConfigSource configSource : config.getConfigSources()) {
+            if (configSource.getName().contains("application.properties")) {
+                Assertions.fail();
+            }
+        }
+    }
+}

--- a/extensions/picocli/runtime/src/main/java/io/quarkus/picocli/runtime/PicocliConfigBuilder.java
+++ b/extensions/picocli/runtime/src/main/java/io/quarkus/picocli/runtime/PicocliConfigBuilder.java
@@ -1,0 +1,34 @@
+package io.quarkus.picocli.runtime;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.microprofile.config.spi.ConfigSourceProvider;
+
+import io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader;
+import io.quarkus.runtime.configuration.ConfigBuilder;
+import io.smallrye.config.DotEnvConfigSourceProvider;
+import io.smallrye.config.SmallRyeConfigBuilder;
+
+public class PicocliConfigBuilder implements ConfigBuilder {
+    @Override
+    public SmallRyeConfigBuilder configBuilder(final SmallRyeConfigBuilder builder) {
+        // Ideally we would create a new source builder, but it requires a lot of code to copy all required stuff, when we just want to remove a few sources
+        List<ConfigSourceProvider> originalProviders = new ArrayList<>(builder.getSourceProviders());
+        for (ConfigSourceProvider originalProvider : originalProviders) {
+            if (originalProvider instanceof ApplicationPropertiesConfigSourceLoader.InClassPath) {
+                builder.getSourceProviders().remove(originalProvider);
+            } else if (originalProvider instanceof ApplicationPropertiesConfigSourceLoader.InFileSystem) {
+                builder.getSourceProviders().remove(originalProvider);
+            } else if (originalProvider instanceof DotEnvConfigSourceProvider) {
+                builder.getSourceProviders().remove(originalProvider);
+            }
+        }
+        return builder;
+    }
+
+    @Override
+    public int priority() {
+        return Integer.MAX_VALUE;
+    }
+}

